File: server/app/matcha/models/workforce_compliance.py
"""Pydantic models for the Workforce Compliance bundle (business-first EPL trackers)."""

from datetime import date, datetime
from typing import Optional, Literal
from uuid import UUID

from pydantic import BaseModel, Field


# --- AI hiring-tool audit register ---

class HiringAiAuditCreate(BaseModel):
    tool_name: str = Field(..., min_length=1, max_length=255)
    vendor: Optional[str] = None
    purpose: Optional[str] = None
    last_audit_date: Optional[date] = None
    cadence_days: int = Field(default=365, ge=1, le=3650)
    notes: Optional[str] = None


class HiringAiAuditUpdate(BaseModel):
    tool_name: Optional[str] = Field(default=None, min_length=1, max_length=255)
    vendor: Optional[str] = None
    purpose: Optional[str] = None
    last_audit_date: Optional[date] = None
    cadence_days: Optional[int] = Field(default=None, ge=1, le=3650)
    notes: Optional[str] = None


class HiringAiAuditResponse(BaseModel):
    id: UUID
    company_id: UUID
    tool_name: str
    vendor: Optional[str] = None
    purpose: Optional[str] = None
    last_audit_date: Optional[date] = None
    cadence_days: int
    next_due_date: Optional[date] = None
    is_overdue: bool
    notes: Optional[str] = None
    created_at: datetime


# --- Biometric / BIPA consent inventory ---

CollectionType = Literal["fingerprint", "face", "iris", "voice", "hand_geometry", "other"]
ConsentMethod = Literal["written", "digital", "verbal", "other"]


class BiometricPointCreate(BaseModel):
    collection_type: CollectionType
    location_id: Optional[UUID] = None
    purpose: Optional[str] = None
    consent_obtained: bool = False
    consent_obtained_date: Optional[date] = None
    consent_method: Optional[ConsentMethod] = None
    retention_policy: Optional[str] = None
    notes: Optional[str] = None


class BiometricPointUpdate(BaseModel):
    collection_type: Optional[CollectionType] = None
    location_id: Optional[UUID] = None
    purpose: Optional[str] = None
    consent_obtained: Optional[bool] = None
    consent_obtained_date: Optional[date] = None
    consent_method: Optional[ConsentMethod] = None
    retention_policy: Optional[str] = None
    is_active: Optional[bool] = None
    notes: Optional[str] = None


class BiometricPointResponse(BaseModel):
    id: UUID
    company_id: UUID
    location_id: Optional[UUID] = None
    collection_type: str
    purpose: Optional[str] = None
    consent_obtained: bool
    consent_obtained_date: Optional[date] = None
    consent_method: Optional[str] = None
    retention_policy: Optional[str] = None
    is_active: bool
    notes: Optional[str] = None
    created_at: datetime


# --- Pay transparency per-state status ---

class PayTransparencyStateRow(BaseModel):
    state: str
    required: bool
    status: str          # compliant | action_needed | na
    postings_include_ranges: bool
    note: Optional[str] = None
    updated_at: Optional[datetime] = None


class PayTransparencyUpdate(BaseModel):
    status: Literal["compliant", "action_needed", "na"]
    postings_include_ranges: bool = False
    note: Optional[str] = None


# --- Pay-equity study register ---

class PayEquityReviewCreate(BaseModel):
    review_date: Optional[date] = None
    scope: Optional[str] = None
    methodology: Optional[str] = None
    gap_pct: Optional[float] = Field(default=None, ge=0, le=100)
    remediation: Optional[str] = None
    cadence_days: int = Field(default=365, ge=1)
    notes: Optional[str] = None


class PayEquityReviewUpdate(BaseModel):
    review_date: Optional[date] = None
    scope: Optional[str] = None
    methodology: Optional[str] = None
    gap_pct: Optional[float] = Field(default=None, ge=0, le=100)
    remediation: Optional[str] = None
    cadence_days: Optional[int] = Field(default=None, ge=1)
    notes: Optional[str] = None

File: client/src/types/workforceCompliance.ts
// Workforce Compliance — business-first EPL risk trackers.

export interface AiAudit {
  id: string
  company_id: string
  tool_name: string
  vendor: string | null
  purpose: string | null
  last_audit_date: string | null
  cadence_days: number
  next_due_date: string | null
  is_overdue: boolean
  notes: string | null
  created_at: string
}

export type CollectionType = 'fingerprint' | 'face' | 'iris' | 'voice' | 'hand_geometry' | 'other'
export type ConsentMethod = 'written' | 'digital' | 'verbal' | 'other'

export interface BiometricPoint {
  id: string
  company_id: string
  location_id: string | null
  collection_type: CollectionType
  purpose: string | null
  consent_obtained: boolean
  consent_obtained_date: string | null
  consent_method: ConsentMethod | null
  retention_policy: string | null
  is_active: boolean
  notes: string | null
  created_at: string
}

export type PayTransparencyStatus = 'compliant' | 'action_needed' | 'na'

export interface PayTransparencyRow {
  state: string
  required: boolean
  status: PayTransparencyStatus
  postings_include_ranges: boolean
  note: string | null
  updated_at: string | null
}

export interface PayEquityReview {
  id: string
  company_id: string
  review_date: string | null
  scope: string | null
  methodology: string | null
  // A measured protected-class pay gap, or null when none was measured. Distinct
  // from dispersion_pct (within-role spread, which seniority can explain) — the two
  // used to share this field, so a "40% gap" was often 40% of roles showing spread.
  gap_pct: number | null
  dispersion_pct: number | null
  remediation: string | null
  cadence_days: number
  next_due_date: string | null
  is_overdue: boolean
  notes: string | null
  created_at: string
}

export type PayEquitySeverity = 'flag' | 'watch' | 'ok'

export interface PayEquityRole {
  title: string
  n: number
  median: number
  min: number
  max: number
  p25: number
  p75: number
  spread_pct: number
  iqr_pct: number
  range_ratio: number | null
  below_band_n: number
  remediation_cost: number
  severity: PayEquitySeverity
}

export type PayEquityPostureBand = 'equitable' | 'watch' | 'action' | 'insufficient'

export interface PayEquityPosture {
  band: PayEquityPostureBand
  label: string
}

// One ranked "fix first" item — the actionable layer over the per-role table.
export interface PayEquityPriorityAction {
  title: string
  severity: PayEquitySeverity
  below_band_n: number
  remediation_cost: number
  spread_pct: number
  action: string
}

// One role's protected-class comparison. Only classes with n ≥ min_class_cell appear;
// smaller cells are counted in suppressed_n and never named.
export interface PayEquityClassGap {
  title: string
  gap_pct: number
  reference: string
  lowest: string
  n: number
  classes: { class: string; n: number; median: number }[]
  suppressed_n: number
}

// Full result of the pay-equity engine (pay_equity_analysis.analyze).
export interface PayEquityAnalysisResult {
  employee_count: number
  analyzed_roles: number
  flagged_roles: number
  headline_gap_pct: number
  // null = not measured (no/insufficient HRIS demographics), never 0 — "we didn't
  // look" and "we looked and found parity" must stay distinguishable.
  class_gap_pct: number | null
  class_gaps: PayEquityClassGap[]
  demographics_coverage_pct: number
  class_gap_measurable: boolean
  min_class_cell: number
  worst: PayEquityRole | null
  roles: PayEquityRole[]
  total_payroll: number
  median_spread_pct: number
  employees_below_band: number
  flagged_payroll_pct: number
  remediation_estimate: number
  band_floor_pct: number
  posture: PayEquityPosture
  priority_actions: PayEquityPriorityAction[]
}

export interface WorkforceSummary {
  ai_audits: { total: number; overdue: number }
  biometric: { active: number; missing_consent: number }
  pay_equity: { total: number; overdue: number }
  pay_transparency: { required_states: number; action_needed: number }
}
